### ISSUE TITLE
RuntimeError: _th_uniform_ not supported on CPUType for Long

### ISSUE THREAD

**Issue by ttbek** [pytorch/pytorch#20070]
## 🐛 Bug

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: _th_uniform_ not supported on CPUType for Long

## To Reproduce
Follow 1st tutorial https://pytorch.org/tutorials/beginner/blitz/tensor_tutorial.html#sphx-glr-beginner-blitz-tensor-tutorial-py
The error occurs when specifying the dtype:  
`x = torch.zeros(5, 3, dtype=torch.long)`

1. Install Nvidia driver
2. Run the docker image nvidia/cuda:9.0-cudnn7-devel
3. Install conda inside the image
4. Use conda to install pytorch: 
    conda install pytorch torchvision cudatoolkit=9.0 -c pytorch
5. Run python and follow the aforementioned tutorial

## Expected behavior
No error, print(x) yielding:  
tensor([[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]])  
According to said tutorial.  

## Environment

(base) root@4624f2f6cd3f:/# python collect_env.py 
Collecting environment information...
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: Could not collect

Python version: 2.7
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti

Nvidia driver version: 390.116
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.5.0

Versions of relevant libraries:
[pip] numpy==1.16.2
[pip] numpydoc==0.8.0
[pip] torch==1.1.0
[pip] torchvision==0.2.2
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.3                      199  
[conda] mkl-service               1.1.2            py27he904b0f_5  
[conda] mkl_fft                   1.0.10           py27ha843d7b_0  
[conda] mkl_random                1.0.2            py27hd81dba3_0  
[conda] pytorch                   1.1.0           py2.7_cuda9.0.176_cudnn7.5.1_0    pytorch
[conda] torchvision               0.2.2                      py_3    pytorch

## Additional context
(base) root@4624f2f6cd3f:/# lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                16
On-line CPU(s) list:   0-15
Thread(s) per core:    2
Core(s) per socket:    8
Socket(s):             1
NUMA node(s):          1
Vendor ID:             AuthenticAMD
CPU family:            23
Model:                 1
Model name:            AMD Ryzen 7 1800X Eight-Core Processor
Stepping:              1
CPU MHz:               2675.483
CPU max MHz:           3600.0000
CPU min MHz:           2200.0000
BogoMIPS:              7199.13
Virtualization:        AMD-V
Hypervisor vendor:     vertical
Virtualization type:   full
L1d cache:             32K
L1i cache:             64K
L2 cache:              512K
L3 cache:              8192K
NUMA node0 CPU(s):     0-15
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate sme ssbd vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 xsaves clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca

**Comment by ttbek** [comment]
Weird... it seems like it is working now.  Well... I'm lost on how to reproduce it then, will update if I find out.

**Comment by soumith** [comment]
reopen the issue when you can reproduced it again :)

**Comment by MaxLiu245** [comment]
Well, I meet this issue, too.

My Win10 supports me ```CUDA 9.1``` so I thought I downloaded the right corresponding ```torch```.

I just opened my Anaconda, created a new environment with Python 3.7, then installed ```torch``` and ```torchvision```, and I tried to run

```x = torch.zeros(3, 3, dtype=torch.long)```,

the it gave me ```RuntimeError: _th_uniform_ not supported on CPUType for Long``` but I didn't know why...

**Comment by CoruNethron** [comment]
Meet the same issue under Win10 x64
Just trying to launch very first examples from very first pytorch tutorial.
If You run `x = torch.rand(5, 3, dtype=torch.long)` instead of `x = torch.zeros(5, 3, dtype=torch.long)` then you will probably got exactly same error.
Figured it out while writing this comment. So `zeros` function accepts dtype argument just fine.

**Comment by BramVanroy** [comment]
I can confirm that passing `torch.long` (or `torch.int64`) to the dtype in `rand` or `randn` throws an error:

`RuntimeError: _th_normal_ not supported on CPUType for Long`

**Comment by phill-barnett** [comment]
You may have been getting this error if you were using a function besides `zeroes`. I imagine many people ended up here for that reason, so just in case: Like @CoruNethron said, the `dtype` argument that you probably grabbed from the [60-Minute Blitz Tutorial](https://pytorch.org/tutorials/beginner/blitz/tensor_tutorial.html#sphx-glr-beginner-blitz-tensor-tutorial-py) is meant for the `zeros` function specifically.
`x = torch.zeros(5, 3, dtype=torch.long)`
If you try to use it with `rand` for instance, you'll get the following error:
`RuntimeError: _th_uniform_ not supported on CPUType for Long`

**Comment by matiu2** [comment]
Summarising, the error message you want to read is most likely:

*"rand needs floating point numbers, (because every value it generates is between 0 and 1), but you're trying to make a 'rand' array of integers"*

**Fix: Stop using Integers with rand.**

(I got here from using [tch-rs](https://github.com/LaurentMazare/tch-rs) with [this tutorial](https://pytorch.org/tutorials/beginner/blitz/tensor_tutorial.html#sphx-glr-beginner-blitz-tensor-tutorial-py)) my code was: `Tensor::rand(&[5, 6], (Kind::Int, Device::Cpu));` and the fix was to change it to `Tensor::rand(&[5, 6], (Kind::Float, Device::Cpu));`

**Comment by ttbek** [comment]
It's semantics, but I don't like calling that the fix as it is not the fix for the bug I described, but for random people ending up here with the same error message, not the same thing.  Granted, I may never be able to reproduce it, but at least myself and @MaxLiu245 had it happen I believe only on our first attempt in a brand new environment but on different platforms.  For zeros.  Subsequent posts probably shouldn't usurp the (admittedly closed) bug report.  

Not that creating a more user friendly error message for that might not be helpful, but it probably needs it's own change request or whatever, not to take over this one.  There is nothing more frustrating than having a problem, finding a thread with a title describing exactly that, and then having it be set aside because they solved some other problem.  I would much rather see 'closed, could not reproduce' than 'closed, we fixed something else.'  At least in the first instance users with the actual same problem might be able to contribute the reproduction or it may become more clear if more people indicate that they are experiencing exactly that.  In the later case it's usually consigned to the garbage bin of fixed and no one will even look at people reporting the original issue because they think of the thread as now being about the other one.  My error came from zeros, not rand.
